File: hackerrank/linked_lists/findMergeNode.java
static int findMergeNode(Node head1, Node head2) {
  Node node1 = head1, node2 = head2;
  
  while (node1 != node2) {
    node1 = node1.next == null ? head2 : node1.next;
    node2 = node2.next == null ? head1 : node2.next;
  }
  
  return node1.data;
}

File: hackerrank/linked_lists/findMergeNode.rb
def findMergeNode(head1, head2)
  seen = {}
  node = head1

  while node
    seen[node] = true
    node = node.next
  end

  node = head2
  
  while node
    return node.data if seen[node]
    node = node.next
  end

  nil
end